### PR TITLE
Channel-wise mean layer

### PR DESCRIPTION
**Pull request by timmoon10**
This PR implements channel-wise mean reduction for data-parallel data, on both CPU and GPU. It is a temporary solution toward #474. At instantiation, the layer flashes a warning that its functionality may disappear at any time. Gradient checks come back clean.

The long-term goal is to replace it with a generic reduce_mean layer that can reduce tensors along arbitrary axes. A reduce_sum layer is the transpose of a broadcast layer, so this is on our pathway.

### CODE REVIEW

File: src/layers/misc/channelwise_mean.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     ::fp_compute() {
+
+  // Local matrices
+  const auto& local_input = get_local_prev_activations();
+  auto& local_output = get_local_activations();
+
+  // Dimensions
+  const auto& input_dims = get_input_dims();
+  const El::Int num_channels = input_dims[0];
+  const El::Int channel_size = std::accumulate(input_dims.begin() + 1,
```
Comment (ndryden) [inline src/layers/misc/channelwise_mean.cpp]: Calling this "channel size" is a bit confusing, and could maybe do with a comment for clarification. What this really is is the size of a single channel of a single sample. (My first intuition was that it was the size of the channel dimension.)

File: src/layers/misc/channelwise_mean.cu [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      DataType private_sum = 0;
+      for (El::Int i = gidx; i < channel_size; i += nthreadsx) {
+        private_sum += input[i + channel*channel_size + col*input_ldim];
+      }
+
+      // Shared memory reduction to get sum for each block
+      /// @todo unroll loops
+      __shared__ DataType shared_sums[block_size];
+      shared_sums[tid] = private_sum;
+      for (El::Int stride = block_size / 2; stride > 0; stride /= 2) {
```
Comment (ndryden) [inline src/layers/misc/channelwise_mean.cu]: This traversal pattern could do with a quick comment, it's not especially clear what's going on.

**Comment by ndryden** [comment]
I agree a generic `reduce_mean` layer would be a nice thing to have. In some sense, what we want is the equivalent of `np.mean` (or `torch.mean`) where you can specify a axis/dimension to reduce over.

File: src/layers/misc/channelwise_mean.cu (hunk 2 above)
Comment (naoyam) [reply src/layers/misc/channelwise_mean.cu]: Cub has some primitives that hide dirty details of reductions.
